pkg/controller/deployment/deployment_controller_test.go

```diff
@@ -34,6 +35,114 @@ import (
 	"k8s.io/kubernetes/pkg/util/intstr"
 )
 
+func TestProportionallyScaling(t *testing.T) {
+	tests := []struct {
+		name       string
+		deployment exp.Deployment
+
+		newRS  *exp.ReplicaSet
+		oldRSs []*exp.ReplicaSet
+
+		expectedNew *exp.ReplicaSet
+		expectedOld []*exp.ReplicaSet
+	}{
+		{
+			name:       "normal scaling event: 10 -> 12",
+			deployment: deployment("foo", 12, intstr.FromInt(1), intstr.FromInt(0)),
+
+			newRS:  rs("foo-v1", 10, nil),
+			oldRSs: []*exp.ReplicaSet{},
+
+			expectedNew: rs("foo-v1", 12, nil),
+			expectedOld: []*exp.ReplicaSet{},
+		},
+		{
+			name:       "normal scaling event: 10 -> 5",
+			deployment: deployment("foo", 5, intstr.FromInt(0), intstr.FromInt(1)),
+
+			newRS:  rs("foo-v1", 10, nil),
+			oldRSs: []*exp.ReplicaSet{},
+
+			expectedNew: rs("foo-v1", 5, nil),
+			expectedOld: []*exp.ReplicaSet{},
+		},
+		{
+			name:       "proportional scaling: 5 -> 10",
+			deployment: deployment("foo", 10, intstr.FromInt(1), intstr.FromInt(0)),
+
+			newRS:  rs("foo-v2", 2, nil),
+			oldRSs: []*exp.ReplicaSet{rs("foo-v1", 3, nil)},
+
+			expectedNew: rs("foo-v2", 4, nil),
+			expectedOld: []*exp.ReplicaSet{rs("foo-v1", 6, nil)},
+		},
+		{
+			name:       "proportional scaling: 5 -> 2",
+			deployment: deployment("foo", 2, intstr.FromInt(0), intstr.FromInt(0)),
+
+			newRS:  rs("foo-v2", 2, nil),
+			oldRSs: []*exp.ReplicaSet{rs("foo-v1", 3, nil)},
+
+			expectedNew: rs("foo-v2", 1, nil),
+			expectedOld: []*exp.ReplicaSet{rs("foo-v1", 1, nil)},
+		},
+		{
+			name:       "proportional scaling: 9 -> 4",
+			deployment: deployment("foo", 4, intstr.FromInt(0), intstr.FromInt(0)),
+
+			newRS:  rs("foo-v2", 8, nil),
+			oldRSs: []*exp.ReplicaSet{rs("foo-v1", 1, nil)},
+
+			expectedNew: rs("foo-v2", 4, nil),
+			expectedOld: []*exp.ReplicaSet{rs("foo-v1", 0, nil)},
+		},
+		{
+			name:       "proportional scaling: 7 -> 10",
+			deployment: deployment("foo", 10, intstr.FromInt(2), intstr.FromInt(0)),
+
+			newRS:  rs("foo-v2", 2, nil),
+			oldRSs: []*exp.ReplicaSet{rs("foo-v1", 3, nil), rs("foo-v1", 2, nil)},
```

Comment:
While rolling-update, the total replicas should be deployment's size + maxSurge, `9` in this case. But the sum of replicas of all RSes is `7` here. 

What if the deployment scale from 7 to 9? Will the test pass?